pkg/util/runner_test.go

```diff
@@ -0,0 +1,47 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"io"
+	"os"
+	"path/filepath"
+	"runtime"
+	"testing"
+)
+
+func Test_Runner_Run(t *testing.T) {
+	var runner Runner = &RealRunner{}
+	var stdOut, stdErr io.Writer
+
+	currentDir, _ := os.Getwd()
+	dummybinary := filepath.Join(currentDir, "..", "..", "test", "testdata", "dummybinary")
```

Comment:
I don't think this is quite right as test. You want to just pass the command name, not the fully specified path. If you do so, you are bypassing the whole lookup  part. 

Pass just the binary name and without additional modifications, it should not work. Then in a second test, adjust the PATH first and then it should work.

BTW, what binaries are we talking about here? What do they do?

Reply:
I think it would be better to have a _testdata_ directory under _pkg/util_ directly. Adding all testdata under the same directory does not really scale so well. For some terst data which is shared between many tests it might make sense, but otherwise I think we should use testdata local to the test. Golang ignores directories called _testdata_ for the actual binary build. It is only used for tests. See also _/pkg/minishift/docker/image/testdata_.

Reply:
> BTW, what binaries are we talking about here? What do they do?

The binaries are dead simple they just print "Minishift", using `fmt.Println()`, just to check the exitcode returned by `Runner`.

Reply:
> Pass just the binary name and without additional modifications, it should not work. Then in a second test, adjust the PATH first and then it should work.

Just to make sure i understand what you mean, you want the _command_ to be directly passed to `Runner`? But `Runner` takes the full path, i did try passing only the _command_ and faced some issue and @praveenkumar helped me to understand that it takes the commnad path and then it converts it to a absolute path.

Reply:
Oh, I see. I guess we are talking about `path, _ := filepath.Abs(commandPath)`. Actually, I would consider this a bug or at least a bad implementation choice. I don't think we need to make an absolute path at this point.  It prevents us from utilizing the full functionality of exec.Command.

Also, for the case we use it, we already have the full path. Mostly we need it for `oc` and there we get the full path from instance config. 

I'd recommend removing this line creating the absolute path.

Reply:
> Just to make sure i understand what you mean, you want the command to be directly passed to Runner? 

Yes

> But Runner takes the full path, it doesn't do a exec.LookPath()

It should. See comment above. Also, check the Godoc of `exec.Command`: "If name contains no path separators, Command uses LookPath to resolve name to a complete path if possible. Otherwise it uses name directly as Path"

Reply:
Thanks @hferentschik will make the change in `Runner`

Reply:
This doesn't seem to help, for some reason `Runner` is not returning the correct exitcode when its called from the unit test function.

However it works fine in the actual Pre-flight check.

Reply:
> This doesn't seem to help, for some reason Runner is not returning the correct exitcode when its called from the unit test function.

So you are removing the line _path, _ := filepath.Abs(commandPath)_, right? And the test is not returning what you are after? What is it returning and what do you expect? 

Note, the test here is not quite the right if you want to test the lookup via the path. You are building an absolute path, so the lookup won't be made. There are several test cases you need to cover.

Reply:
> So you are removing the line path, _ := filepath.Abs(commandPath), right? And the test is not returning what you are after? What is it returning and what do you expect?

Yes and yes, its not returning the correct exit code, returns 1 every time.

> Note, the test here is not quite the right if you want to test the lookup via the path. You are building an absolute path, so the lookup won't be made.

Yes and that is the issue, the _exec.Command()_ and _exec.LookPath()_ is behaving weired in case of the unit tests.

Reply:
https://github.com/anjannath/minishift/pull/2

Reply:
The changes in the PR [anjannath#2](https://github.com/anjannath/minishift/pull/2) fixed the issue, i was using _ls_ and _dir_ for the command lookup test, in place of adding the dummybinaries to the PATH, and that wasn't working.